File: common/src/main/java/juuxel/adorn/client/book/Image.java
package juuxel.adorn.client.book;

import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import juuxel.adorn.util.Vec2i;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.ComponentSerialization;
import net.minecraft.resources.Identifier;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public record Image(Identifier location, Vec2i size, Placement placement, List<HoverArea> hoverAreas) {
    public static final Codec<Image> CODEC = RecordCodecBuilder.create(instance -> instance.group(
        Identifier.CODEC.fieldOf("location").forGetter(Image::location),
        Vec2i.CODEC.fieldOf("size").forGetter(Image::size),
        Placement.CODEC.optionalFieldOf("placement", Placement.AFTER_TEXT).forGetter(Image::placement),
        HoverArea.CODEC.listOf().optionalFieldOf("hoverAreas", List.of()).forGetter(Image::hoverAreas)
    ).apply(instance, Image::new));

    public enum Placement {
        BEFORE_TEXT("beforeText"),
        AFTER_TEXT("afterText");

        private static final Map<String, Placement> BY_ID = Arrays.stream(values())
            .map(placement -> Pair.of(placement.id, placement))
            .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond));
        public static final Codec<Placement> CODEC = Codec.STRING.xmap(BY_ID::get, placement -> placement.id);

        private final String id;

        Placement(String id) {
            this.id = id;
        }
    }

    public record HoverArea(Vec2i position, Vec2i size, Component tooltip) {
        public static final Codec<HoverArea> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            Vec2i.CODEC.fieldOf("position").forGetter(HoverArea::position),
            Vec2i.CODEC.fieldOf("size").forGetter(HoverArea::size),
            ComponentSerialization.CODEC.fieldOf("tooltip").forGetter(HoverArea::tooltip)
        ).apply(instance, HoverArea::new));

        public boolean contains(int x, int y) {
            return position.x() <= x && x <= position.x() + size.x() && position.y() <= y && y <= position.y() + size.y();
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/book/Image.kt
package juuxel.adorn.client.book

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import juuxel.adorn.util.MoreCodecs
import juuxel.adorn.util.Vec2i
import net.minecraft.text.Text
import net.minecraft.util.Identifier

data class Image(
    val location: Identifier,
    val size: Vec2i,
    val placement: Placement,
    val hoverAreas: List<HoverArea>
) {
    companion object {
        val CODEC: Codec<Image> = RecordCodecBuilder.create { instance ->
            instance.group(
                Identifier.CODEC.fieldOf("location").forGetter { it.location },
                Vec2i.CODEC.fieldOf("size").forGetter { it.size },
                Placement.CODEC.optionalFieldOf("placement", Placement.AFTER_TEXT).forGetter { it.placement },
                HoverArea.CODEC.listOf().optionalFieldOf("hoverAreas", emptyList()).forGetter { it.hoverAreas }
            ).apply(instance, ::Image)
        }
    }

    enum class Placement(private val id: String) {
        BEFORE_TEXT("beforeText"),
        AFTER_TEXT("afterText");

        companion object {
            private val BY_ID: Map<String, Placement> = values().associateBy { it.id }
            val CODEC: Codec<Placement> = Codec.STRING.xmap(BY_ID::get, Placement::id)
        }
    }

    data class HoverArea(val position: Vec2i, val size: Vec2i, val tooltip: Text) {
        fun contains(x: Int, y: Int): Boolean =
            x in position.x..(position.x + size.x) && y in position.y..(position.y + size.y)

        companion object {
            val CODEC: Codec<HoverArea> = RecordCodecBuilder.create { instance ->
                instance.group(
                    Vec2i.CODEC.fieldOf("position").forGetter { it.position },
                    Vec2i.CODEC.fieldOf("size").forGetter { it.size },
                    MoreCodecs.TEXT.fieldOf("tooltip").forGetter { it.tooltip }
                ).apply(instance, ::HoverArea)
            }
        }
    }
}
